homeassistant/components/plum_lightpad/__init__.py

```diff
@@ -26,56 +28,53 @@
     extra=vol.ALLOW_EXTRA,
 )
 
+PLATFORMS = ["light"]
 
-async def async_setup(hass, config):
+
+async def async_setup(hass: HomeAssistant, config: dict):
     """Plum Lightpad Platform initialization."""
+    if DOMAIN not in config:
+        return True
 
     conf = config[DOMAIN]
-    plum = Plum(conf[CONF_USERNAME], conf[CONF_PASSWORD])
 
-    hass.data[DOMAIN] = plum
+    _LOGGER.info("Found Plum Lightpad configuration in config, importing...")
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+        )
+    )
 
-    def cleanup(event):
-        """Clean up resources."""
-        plum.cleanup()
+    return True
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, cleanup)
 
-    cloud_web_sesison = async_get_clientsession(hass, verify_ssl=True)
-    await plum.loadCloudData(cloud_web_sesison)
-
-    async def new_load(device):
-        """Load light and sensor platforms when LogicalLoad is detected."""
-        await asyncio.wait(
-            [
-                hass.async_create_task(
-                    discovery.async_load_platform(
-                        hass, "light", DOMAIN, discovered=device, hass_config=conf
-                    )
-                )
-            ]
-        )
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Plum Lightpad from a config entry."""
+    _LOGGER.debug("Setting up config entry with ID = %s", entry.unique_id)
 
-    async def new_lightpad(device):
-        """Load light and binary sensor platforms when Lightpad detected."""
-        await asyncio.wait(
-            [
-                hass.async_create_task(
-                    discovery.async_load_platform(
-                        hass, "light", DOMAIN, discovered=device, hass_config=conf
-                    )
-                )
-            ]
-        )
+    username = entry.data.get(CONF_USERNAME)
+    password = entry.data.get(CONF_PASSWORD)
 
-    device_web_session = async_get_clientsession(hass, verify_ssl=False)
-    hass.async_create_task(
-        plum.discover(
-            hass.loop,
-            loadListener=new_load,
-            lightpadListener=new_lightpad,
-            websession=device_web_session,
+    try:
+        plum = await load_plum(username, password, hass)
+    except ContentTypeError as ex:
+        _LOGGER.error("Unable to authenticate to Plum cloud: %s", ex)
+        return False
+    except (ConnectTimeout, HTTPError) as ex:
+        _LOGGER.error("Unable to connect to Plum cloud: %s", ex)
+        raise ConfigEntryNotReady
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = plum
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
         )
-    )
 
+    def cleanup(event):
+        """Clean up resources."""
+        plum.cleanup()
```

Comment:
Should this run in the event loop or in the thread pool?

Reply:
I don't know. Happy to make the change if you could point me to some 'best practices' / 'architecture principles'  or similar docs - just want to make sure I understand what I am doing

Reply:
Currently it will run in the thread pool since the callback `cleanup` isn't decorated with `@callback`.

I looked in the library and this callback joins a thread, so it should not run in the event loop so that's good.

I'd add a code comment about this as the library is mixing async and sync context in its api.

Reply:
@MartinHjelmare We can take on fixing the underlying library too. What would you like changed?

Currently each [lightpad starts it's own thread](https://github.com/ColinHarrington/plum-lightpad-python/blob/master/plumlightpad/lightpad.py#L22-L25) which receives change events (motion, power, config, etc.) currently using `telnetlib` and `threading`.

The [`plum.cleanup()`](https://github.com/ColinHarrington/plum-lightpad-python/blob/master/plumlightpad/plumlightpad.py#L91-L93) function [shuts down those threads](https://github.com/ColinHarrington/plum-lightpad-python/blob/master/plumlightpad/lightpad.py#L59-L61).

I'd like to have those open tcp connections asyncio driven.  What's the desired model here?  Is it expected that underlying lib would start it's own event loop or join home assistant's main loop?

The discovery process/UDP endpoint is created in the event loop passed in (: https://github.com/ColinHarrington/plum-lightpad-python/blob/master/plumlightpad/plumlightpad.py#L59-L71

Reply:
The library should use the same event loop as home assistant.